src/client/protocol.c

```diff
@@ -0,0 +1,832 @@
+#include <inttypes.h>
+#include <poll.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <unistd.h>
+
+#include "../lib/assert.h"
+
+#include "protocol.h"
+#include "../message.h"
+#include "../protocol.h"
+#include "../request.h"
+#include "../response.h"
+#include "../tracing.h"
+#include "../tuple.h"
+
+/* Read data from fd into buf until one of the following occurs:
+ *
+ * - The full count n of bytes is read.
+ * - A read returns 0 (EOF).
+ * - Polling the fd in preparation for a read times out.
+ *
+ * On error, -1 is returned. Otherwise the return value is the count
+ * of bytes read. This may be less than n if either EOF happened or
+ * polling timed out. */
+static ssize_t doRead(int fd, void *buf, size_t n, int timeout_millis)
+{
+	size_t got = 0;
+	struct pollfd pfd;
+	ssize_t rv;
+
+	pfd.fd = fd;
+	pfd.events = POLLIN;
+	pfd.revents = 0;
+
+	while (got < n) {
+		rv = poll(&pfd, 1, timeout_millis);
+		if (rv < 0) {
+			if (errno == EINTR) {
+				continue;
+			} else {
+				return rv;
+			}
+		} else if (rv == 0) {
+			// timeout
+			// XXX signal this to the caller somehow?
+			break;
+		}
+		assert(rv == 1);
+		if (pfd.revents != POLLIN) {
+			return -1;
+		}
+		rv = read(fd, (char *)buf + got, n - got);
+		if (rv < 0) {
+			return rv;
+		} else if (rv == 0) {
+			break;
+		}
+		got += (size_t)rv;
+	}
+	return (ssize_t)got;
+}
+
+int clientInit(struct client_proto *c,
+	       int fd,
+	       int timeout_first_millis,
+	       int timeout_follow_millis)
+{
+	tracef("init client");
+	int rv;
+	c->fd = fd;
+	c->db_name = NULL;
+	c->db_is_init = false;
+
+	c->timeout_first_millis = timeout_first_millis;
```

Comment:
Why does the distinction between `timeout_first_millis` and `timeout_follow_millis` exist?

Reply:
Yes, something like that could work I think. Same then for `Send`?